### ISSUE TITLE
LDAP Setup

### ISSUE THREAD

**Issue by kavitha0624** [Hygieia/Hygieia#1439]
Hi Team,

I took the latest code from the Master branch and trying to setup LDAP. I have added the properties as given below. I do not see any errors in the api console but when I try to login thru LDAP, I get an error msg saying Invalid username and password.

API Properties File:

dbname=dashboardfive
dbusername=dashboarduser
dbpassword=dbpassword
dbhost=localhost
dbport=27017
dbreplicaset=false
dbhostport=localhost:27017
server.contextPath=/api
server.port=8080

#LDAP Server Url, including port of your LDAP server
#auth.ldapServerUrl=[ldap://company.com:389]
auth.ldapServerUrl=ldap://vds.xxx.com:389

#If using standard ldap
#LDAP User Dn Pattern, where the username is replaced with '{0}'
#auth.ldapUserDnPattern=[uid={0},OU=Users,dc=your,dc=company,dc=com]
auth.ldapUserDnPattern=uid={0},OU=people,dc=xxx,dc=com

API Console:

The entire log is below but I found this line which says Dn is empty. Is that is the reason, can you please let me know what the mistake is - [localhost-startStop-1] INFO org.springframework.security.ldap.DefaultSpringSecurityContextSource -  URL 'ldap://vds.xxx.com:389', root DN is ''

Complete Log:

[api_ldap.txt](https://github.com/capitalone/Hygieia/files/1042843/api_ldap.txt)

**Comment by kavitha0624** [comment]
Hi Team,

Can anyone please help me with this issue

**Comment by satishc1** [comment]
@kavitha0624 Are you using AD?

**Comment by kavitha0624** [comment]
@satishc1 I did not get you.. what is AD?

**Comment by kavitha0624** [comment]
Hi @satishc1 can you please let me know what did you mean by AD. This issue is blocking me. Please help.

**Comment by satishc1** [comment]
@kavitha0624 Active Directory. The api properties for AD are different from the properties for LDAP, so if you are using AD, then make sure you set those.

**Comment by kavitha0624** [comment]
@satishc1 , we are using LDAP here. So I have set the properties according to Hygieia setup document.

**Comment by kavitha0624** [comment]
@satishc1 , could you please suggest how do I proceed with this configuration. As you asked, we are using LDAP here. So I am guessing the properties should be correct.

**Comment by rob-miller-777** [comment]
The logs won't show the dn string you give it in the properties file, only the url.  Have you verified that user dn pattern and url outside of Hygieia to make sure it works?

**Comment by rob-miller-777** [comment]
My best guess from what I see here is that the properties are getting through fine, but they are most likely incorrect for your ldap environment.

**Comment by kavitha0624** [comment]
Thanks Rob. Will check once again and update.

**Comment by MarkRx** [comment]
I am reviewing the LDAP code and have found several issues. I'm investigating it now.

**Comment by Sujaydevops** [comment]
any updates here?

**Comment by MarkRx** [comment]
No I have had to work on other things and have not had time to continue my investigation. A few things to note -

- The current implementation doesn't support LDAP binding password correctly (it is only used in the api and not the spring template)
- The current implementation lacks flexibility (i.e. custom filter)
- There are two competing LDAP libraries being used

**Comment by srujan5** [comment]
I'm also facing same issues with Ldap, while running api, it was not even reading Ldap properties. please help me

**Comment by srujan5** [comment]
@kavitha0624  did you resolved that issue, I'm also getting similar error.

**Comment by kavitha0624** [comment]
@srujan5 I could not resolve this issue.. it's still there.

**Comment by justinzyw** [comment]
not sure how to make the LDAP binding user and password to work... I tried to set AUTH_LDAP_BIND_USER and AUTH_LDAP_BIND_PASS but they do not seem working

**Comment by rkatti** [comment]
Anyone able to use LDAP? It's still not working.

**Comment by gars01** [comment]
I follow the rules to configure the properties file for LDAP as tutorial says, but in my case i've got the following error:

Caused by: javax.naming.NameNotFoundException: [LDAP: error code 32 - No Such Object]

Any idea? 
Thanks in advance.

**Comment by matthewbeasley** [comment]
@gars01 can you post your properties file?

**Comment by gars01** [comment]
This is the part about LDAP (I've changed our domain with x's for privacity):

#LDAP Server Url, including port of your LDAP server
#auth.ldapServerUrl=[ldap://company.com:389]
auth.authenticationProviders=LDAP,STANDARD
auth.ldapServerUrl=ldap://ldap-pre.xxxx.es:389

#If using standard ldap
#LDAP User Dn Pattern, where the username is replaced with '{0}'
#auth.ldapUserDnPattern=[uid={0},OU=Users,dc=your,dc=company,dc=com]
auth.ldapUserDnPattern=uid={0},o=xxxxx xx xxxxxxxx-xx xxxxxx,c=ES

Also, our LDAP does not have any permissions logic specified, all users have the same type of access. Thanks in advance.

**Comment by vikramgudi89** [comment]
@gars01 
does LDAP integration with Hygieia is working?

**Comment by gars01** [comment]
Yes and no, I explain, I have traced the code and I have verified that the identification against LDAP is correct, however, the code tries to look for the permissions of the user and it is when it fails. Keep in mind that in my organization we do not have permissions defined in LDAP. When I trace, I see that the value of the DN attribute is null, I do not know if this is significant, nor how to assign a correct value to it, but there is no parameter to do so.
In short, the result is that no error response appears, Hygieia just does not do anything.
